spyder_kernels/comms/commbase.py

```diff
@@ -0,0 +1,527 @@
+        self._reply_inbox = {}
+        self._reply_waitlist = {}
+        self._pickle_protocol = DEFAULT_PICKLE_PROTOCOL
+
+        self._register_message_handler(
+            'remote_call', self._handle_remote_call)
+        self._register_message_handler(
+            'remote_call_reply', self._handle_remote_call_reply)
+
+        # Dummy functions for testing and to trigger side effects such as
+        # an interruption or waiting for a reply.
+        def pong_back():
+            self.remote_call(self.calling_comm_id).pong()
+
+        self.register_call_handler('ping', pong_back)
+        self.register_call_handler('pong', lambda: None)
+        self.register_call_handler('_set_pickle_protocol',
+                                   self._set_pickle_protocol)
+
+    def close(self, comm_id=None):
+        """Close the comm and notify the other side."""
+        if comm_id is None:
+            # close all the comms
+            id_list = list(self._comms.keys())
+        else:
+            id_list = [comm_id]
+
+        for comm_id in id_list:
+            self._comms[comm_id].close()
+            del self._comms[comm_id]
+
+    def is_open(self, comm_id=None):
+        """Check to see if the comm is open."""
+        if comm_id is None:
+            return len(self._comms) > 0
+        return comm_id in self._comms
+
+    def register_call_handler(self, call_name, handler):
+        """
+        Register a remote call handler.
+
+        Parameters
+        ----------
+        call_name : str
+            The name of the called function.
+        handler : callback
+            A function to handle the request, or `None` to unregister
+            `call_name`.
+        """
+        if not handler:
+            self._remote_call_handlers.pop(call_name, None)
+            return
+
+        self._remote_call_handlers[call_name] = handler
+
+    def remote_call(self, comm_id=None, callback=None, **settings):
+        """Get a handler for remote calls."""
+        return RemoteCallFactory(self, comm_id, callback, **settings)
+
+    # ---- Private -----
+    def _send_message(self, spyder_msg_type, content=None, data=None,
+                      comm_id=None):
+        """
+        Publish custom messages to the other side.
+
+        Parameters
+        ----------
+        spyder_msg_type: str
+            The spyder message type
+        content: dict
+            The (JSONable) content of the message
+        data: any
+            Any object that is serializable by cloudpickle (should be most
+            things). Will arrive as cloudpickled bytes in `.buffers[0]`.
+        comm_id: int
+            the comm to send to. If None sends to all comms.
+        """
+        if not self.is_open(comm_id):
+            raise CommError("The comm is not connected.")
+        msg_dict = {
+            'spyder_msg_type': spyder_msg_type,
+            'content': content,
+            'pickle protocol': self._pickle_protocol,
```

Comment:
Please remove the space in this key, i.e. rename it to `pickle_protocol`.

Reply:
Besides, please keep passing a key called `PY2` set to `True` if this is in Python 2 version. We could need that to make other adjustments in the future. Thanks!